### ISSUE TITLE
spring-session-data-redis-2.0.4 produces WARN messages when using multiple redis databases

### ISSUE THREAD

**Issue by joevalerio** [spring-projects/spring-session#1128]
spring-session-data-redis-2.0.4 produces WARN messages when using multiple redis databases

We are using spring-session-data-redis-2.0.4, as part of spring-boot 2.0.3, on redis 2.x,3.x,4.x.  For our development servers we are using different redis databases per instance for our session caches.  

In spring-session-data-redis-1.x the warning message is not present in the RedisOperationsSessionRepository.onMessage method.  This was added in 2.+.

Since messages are not filtered per db we get all key event messages regardless of the database.  The listener that is constructed listens for ALL del, expire, and create messages.

What is troubling here, is for events from other dbs spring-session is throwing SessionCreatedEvents with null session objects, which is not correct.  For the del/expire case, the warning message is logged, but the events are not created, which I think is less wrong, but still wrong.

When registering the listener in RedisHttpSessionConfiguration, we CAN filter on the database for del/expire messages if we add the database in this configuration, but I don't see a way to filter on the create messages, which could be problematic.

For now I am just setting logging to ERROR for this class, to avoid filling my disks, because we don't use these events, and its a non issue, for us.

Repro Steps:
Connect app1 to redis db0
Connect app2 to redis db1
create and delete a session on app1
notice WARN message in app2

NOTE: make sure key events are enabled.

Thanks in advance,

- Joe

**Comment by joevalerio** [comment]
Ok, after debugging some more, when using boot 1.3.8, and spring-session-data-redis-1.2.2 the filtering is not the issue, it never was, the 1.2.2 implementation does not contain the WARN message when failing to find the session for the delete keyevents sessionId.  

This was added. 

RedisOperationsSessionRepository @line 519
```
			if (session == null) {
				logger.warn("Unable to publish SessionDestroyedEvent for session "
						+ sessionId);
				return;
			}

```

Sorry should have seen this yesterday, but I missed it due to history not maintained due to project reorg.

**Comment by vpavic** [comment]
Thanks for the report @joevalerio, that's an interesting problem. I'll try to reproduce locally.

**Comment by vpavic** [comment]
I've reproduced this problem, and have a WIP fix in place locally for handling both create and del/expire messages. The only issue is how to obtain the information about database index in use. ATM I'm supplying it manually, however this isn't ideal as users have already specified the database index in their `RedisConnectionFactory` configuration and would have to configure it in two places.

I don't see a proper way to obtain this information from `RedisConnectionFactory` - @mp911de is there anything you could suggest here?

**Comment by mp911de** [comment]
We're exposing the desired database index through `JedisConnectionFactory` and `LettuceConnectionFactory`. `RedisConnectionFactory` exposes only connection-creating methods. Please note, that the database index on the factory is the one to pre-initialize a new connection. Users might change the database by calling `select(…)` on the connection.

There is no possibility to determine from a connection which database index is selected.

**Comment by vpavic** [comment]
@joevalerio This is fixed in `master` and `2.0.x` - if you could give `2.1.0-BUILD-SNAPSHOT` or `2.0.7-BUILD-SNAPSHOT` a try that would be appreciated.
